logrotate.c

```diff
@@ -1388,6 +1388,111 @@ static long daysElapsed(const struct tm *now, const struct tm *last)
     return (long) ((intmax_t)diff / DAY_SECONDS);
 }
 
+static int isSecureDir(const char *pathname, uid_t exempt_uid, gid_t exempt_gid, int depth);
+
+static int isSecureDirStep(const char *pathname, uid_t exempt_uid, gid_t exempt_gid, int depth)
+{
+    struct stat sb;
+
+    if (lstat(pathname, &sb) < 0)
+        return -errno;
+
+    if (S_ISLNK(sb.st_mode)) {
+        /* check whether symlink target is inside a secure directory */
+        char *canonpath;
+        int r;
+
+        canonpath = realpath(pathname, NULL);
+        if (!canonpath)
+            return -errno;
+
+        r = isSecureDir(canonpath, exempt_uid, exempt_gid, depth + 1);
+        free(canonpath);
+        return r;
+    }
+
+    if (!S_ISDIR(sb.st_mode))
+        return -ENOTDIR;
+
+    if ((sb.st_uid != ROOT_UID && sb.st_uid != exempt_uid) ||
+        ((sb.st_mode & S_IWGRP) && sb.st_gid != 0 && sb.st_gid != exempt_gid) ||
+        (sb.st_mode & S_IWOTH && !(sb.st_mode & S_ISVTX))) {
+        return 0;
+    }
+
+    return 1;
+}
+
+static int isSecureDir(const char *pathname, uid_t exempt_uid, gid_t exempt_gid, int depth)
+{
+    char *copy, *p;
+    int r;
+
+    if (depth >= 10)
+        return -ELOOP;
+
+    if (pathname[0] != '/')
+        return -EINVAL;
+
+    copy = strdup(pathname);
+    if (!copy)
+        return -errno;
+
+    for (p = copy + 1; p; p = strchr(p + 1, '/')) {
+        char backup;
+
+        backup = *p;
+        *p = '\0';
+        r = isSecureDirStep(copy, exempt_uid, exempt_gid, depth);
+        *p = backup;
+        if (r <= 0)
+            goto cleanup;
+    }
+
+    r = isSecureDirStep(pathname, exempt_uid, exempt_gid, depth);
+
+cleanup:
+    free(copy);
+    return r;
+}
+
+/*
+ * Check whether the given path is a secure directory.
+ *
+ * Return 1 if the check passes,
+ *        0 if the check fails or
+ *        <0 negative errno style value on error.
+ */
+int checkIsSecureDir(const char *pathname, const struct logInfo *log)
+{
+    static char lastSecureDir[PATH_MAX];
+    size_t pathlen;
+    uid_t exempt_uid;
+    gid_t exempt_gid;
+    int r;
+
+    if (log->flags & LOG_FLAG_ALLOWNONSECUREDIR)
+        return 1;
+
+    pathlen = strlen(pathname);
+    if (strncmp(lastSecureDir, pathname, pathlen) == 0 &&
+        (lastSecureDir[pathlen] == '\0' || lastSecureDir[pathlen] == '/'))
+        return 1;
+
+    if (getuid() != ROOT_UID && getgid() != 0)
+        return 1;
+
+    exempt_uid = (log->flags & LOG_FLAG_SU) ? log->suUid : ROOT_UID;
+    exempt_gid = (log->flags & LOG_FLAG_SU) ? log->suGid : 0;
+
+    r = isSecureDir(pathname, exempt_uid, exempt_gid, 0);
+    if (r == 1 && pathlen < sizeof(lastSecureDir)) {
```

Comment:
We record the path that was checked successfully the last time.  But cannot it happen that `exempt_uid` or `exempt_gid` will be different in a subsequent call of this function for the same path?

Reply:
true, dropped the LRU cache